Ported car_sim_lcm to System 2.0

liangfok opened the pull request:
This is a version of `car_sim_lcm` that is based on System 2.0. It replaces the original one, which was based on System 1.0.

Resolves #3258.

<!-- Reviewable:start -->
---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/3886)

<!-- Reviewable:end -->

liangfok commented:
I marked this PR "do not review" because there are a few other PRs like #3828 that I would like merged before this one.

jamiesnape commented:
@drake-jenkins-bot linux-gcc-experimental-ros please

jamiesnape commented:
@drake-jenkins-bot linux-cpplint-experimental please
@drake-jenkins-bot linux-gcc-experimental-matlab please

jamiesnape commented:
@drake-jenkins-bot retest this please
@drake-jenkins-bot linux-clang-experimental please
@drake-jenkins-bot linux-clang-experimental-minimal please
@drake-jenkins-bot linux-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-cpplint-experimental please
@drake-jenkins-bot linux-gcc-experimental-matlab please
@drake-jenkins-bot linux-gcc-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-ros please
@drake-jenkins-bot linux-gcc-ninja-experimental please

jamiesnape commented:
@drake-jenkins-bot linux-clang-experimental please
@drake-jenkins-bot linux-clang-experimental-debug please
@drake-jenkins-bot linux-clang-ninja-experimental-minimal please
@drake-jenkins-bot linux-clang-ninja-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-open-source please
@drake-jenkins-bot linux-gcc-experimental-ros please
@drake-jenkins-bot linux-gcc-ninja-experimental please
@drake-jenkins-bot linux-gcc-ninja-experimental-open-source-debug please

liangfok commented:
+@jwnimmer-tri for feature review please.
+@sammy-tri for platform review please.

Thanks!

---

Review status: 0 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-:-KVXA8twP7hipn8_JsKb:bjege30)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Exciting!  I will be sure to take a look this evening.

jwnimmer-tri commented:
Partial checkpoint.

---

Reviewed 5 of 6 files at r1.
Review status: 5 of 6 files reviewed at latest revision, 4 unresolved discussions.

---

*[drake/automotive/automotive_common.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXD9LzxaTj70lEy57Y:-KVXD9LzxaTj70lEy57Z:b-qo5fv8) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/automotive_common.cc#L12)):*
> ```cpp
>   Eigen::Vector4d color;
>   color << 0.9297, 0.7930, 0.6758,
>       1;  // was hex2dec({'ee','cb','ad'})'/256 in matlab
> ```

BTW I'm not sure this comment is adding value anymore?

---

*[drake/automotive/automotive_common.h, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXCdm8ApxD0qF9-N0H:-KVXCdm8ApxD0qF9-N0I:b-x2cu3n) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/automotive_common.h#L26)):*
> ```objc
>  * axis. Note that regardless of how deep the terrain is, the top surface of the
>  * terrain will be at Z = 0.
>  */
> ```

BTW I think this entire overview would be clearer if everything after the first paragraph was replaced with "Two opposite corners of the resulting axis-aligned box are (size/2,size/2,0) and (-size/2,-size/,-depth)" or whatever they are.  I found it difficult to try to parse out the long-form explanation.

---

*[drake/automotive/automotive_common.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXDMQzULyVl38RSuqO:-KVXDMQzULyVl38RSuqP:b-9udt1b) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/automotive_common.h#L31)):*
> ```objc
>                            double box_size = 1000, double box_depth = 10);
> 
> }  // namespace automotive
> ```

BTW I think having a pile of free functions in the module-global namespace will _eventually_ be difficult to understand / maintain, but for now it's a fine step along the path to a larger codebase.  Once this file has a handful of functions and users, we can work on refactoring it given the evidence of whatever needs it ends up meeting.

---

*[drake/automotive/car_simulation.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXE5E0b0_qlAlp0Lvn:-KVXE5E12c2nP-u_Fw5B:b-ot6y4d) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_simulation.cc#L154)):*
> ```cpp
>       "terrain");
>   tree->updateStaticCollisionElements();
> }
> ```

It looks like this method duplicates the same method in automotive_common?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-:-KVXEDQqGz3__PUK2X0S:b5f2erm)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 6 files at r1.
Review status: all files reviewed at latest revision, 19 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXF6_UHTv70kxHkh_e:-KVXF6_V4YZ7lN9f9Enr:b-2lktll) (no related file):*
(N.B. New discussion thread, don't post PR workflow discussions here.)

It seems to me like README.md is out of date for this revised program?  I think its fine to _remove_ the incorrect instructions rather than fix them, but I don't think we should post instructions that don't work.  In particular, I think the list of SDFs on the command-line is no longer supported?

---

*[drake/automotive/car_sim_lcm.cc, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXG_M2nY6rLHJEoQta:-KVXG_M3WQx6sR7fyijz:baik2j3) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L46)):*
> ```cpp
>   DiagramBuilder<double> builder;
> 
>   // Instantiates an Multibody Dynamics (MBD) model of the world.
> ```

BTW I'm not entirely sure we should be using this term (MBD) until the referenced code actually does too.  Whatever, I guess it's just a comment.

---

*[drake/automotive/car_sim_lcm.cc, line 60 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXGyxrjR8itMjtdH9K:-KVXGyxrjR8itMjtdH9L:b-mnzvrg) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L60)):*
> ```cpp
> 
>   // Instantiates a PID controller for controlling the actuators in the
>   // RigidBodyPlant. The vector order is [steering, left wheel, right wheel].
> ```

Please document the units for each of the three items.

---

*[drake/automotive/car_sim_lcm.cc, line 78 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXHBSN4EYmgjAoT-Be:-KVXHBSN4EYmgjAoT-Bf:b86i5px) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L78)):*
> ```cpp
>   //     3   |   steering angle velocity
>   //     4   |   left wheel velocity
>   //     5   |   right wheel velocity
> ```

Please document the units for each of the above.

---

*[drake/automotive/car_sim_lcm.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXHRVD-Hhwocb7A11k:-KVXHRVD-Hhwocb7A11l:b-ncgu05) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L124)):*
> ```cpp
>   const int kStateIndexSteeringAngleVelocity = 20;
>   const int kStateIndexLeftWheelVelocity     = 22;
>   const int kStateIndexRightWheelVelocity    = 24;
> ```

Is this truly the best way to read out the positions and velocities from the tree?  Can't we look up a RigidBody and as it its index?  Or at worst, DEMAND that these hard-coded indicies match up with string Body names from the tree?

---

*[drake/automotive/car_sim_lcm.cc, line 149 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXI8emcGanrds9_tGm:-KVXI8ens8he0DAZ5uVE:b-stviyw) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L149)):*
> ```cpp
>           get_rigid_body_tree(), &lcm);
> 
>   lcm.StartReceiveThread();
> ```

It seems ill-advised to start executing our Systems before the DiagramBuilder::Build call happens, at least?

---

*[drake/automotive/car_sim_lcm.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXIrC8eQEPpCJivW11:-KVXIrC8eQEPpCJivW12:bq8jnj8) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L154)):*
> ```cpp
>   // consists of the following three-vector:
>   //
>   // [steering angle position, throttle velocity, brake velocity]
> ```

The term "throttle velocity" is a bit ambiguous?  Velocity with respect to what position?  Pedal displacement?  Wheel rotation?  Longitudinal vehicle position?

---

*[drake/automotive/car_sim_lcm.cc, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXJCsU8bIqDTTcS-49:-KVXJCsVxB7BVGPiU1l3:b-sk798t) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L156)):*
> ```cpp
>   // [steering angle position, throttle velocity, brake velocity]
>   //
>   static const DrivingCommandTranslator driving_command_translator;
> ```

Remove the word `static`.
https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables

---

*[drake/automotive/car_sim_lcm.cc, line 179 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXKdCtzQa90svxjku9:-KVXKdCtzQa90svxjkuA:bcdk8ir) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L179)):*
> ```cpp
>   // The user's steering angle position command can be passed straight
>   // through. The user's throttle and brake commands need to be multipled by
>   // a gain of 20 and -20 to get the reference velocities for the left and right
> ```

Why 20?  What are the units?

---

*[drake/automotive/car_sim_lcm.cc, line 200 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXKjP23XCNNmBJ6mQb:-KVXKjP23XCNNmBJ6mQc:bgdysx5) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L200)):*
> ```cpp
>       0,  0,   0,
>       0, 20, -20,
>       0, 20, -20;
> ```

To be more obviously failsafe, assert that the matrix size declared by `mimo_gain` was exactly populated by the above left-shift operator.  (Or does Eigen do that automatically?)

---

*[drake/automotive/car_sim_lcm.cc, line 202 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXJU7rIDqiK86Ct3vs:-KVXJU7sivRmONjGMpT0:b5xc6qu) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L202)):*
> ```cpp
>       0, 20, -20;
>   auto user_to_actuator_cmd_sys =
>       builder.template AddSystem<MatrixGain<double>>(mimo_gain);
> ```

BTW This is an acceptable approach.  Another, possibly-simpler approach is writing a small custom direct-feedthrough System that takes in `Value<lcmt_driving_command_t>` on its single abstract-valued port and during its `EvalOutput` produces whatever controller `BasicVector` is desired, using C++ assignments and elementary math.  The Gain matrix abstraction isn't fundamentally necessary.  This could turn the giant comment matrix above into named getters and setters, with immediately-relevant units and scale comments.

---

*[drake/automotive/car_sim_lcm.cc, line 213 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXLKarxX2m6dLZW7rE:-KVXLKascQ6u2eEOUjIC:b74bsve) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L213)):*
> ```cpp
>   // Connects the feedforward torque command.
>   builder.Connect(constant_zero_source->get_output_port(),
>                   controller->get_input_port(0));
> ```

BTW Does the controller offer named port getters, or at least kConstant for these magic 0's and 1's here?

---

*[drake/automotive/car_sim_lcm.cc, line 235 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXLyck7T7lWEpHyp0I:-KVXLyck7T7lWEpHyp0J:b-i9egz0) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L235)):*
> ```cpp
>       diagram->GetMutableSubsystemContext(
>           simulator.get_mutable_context(), controller);
>   controller->SetDefaultState(controller_context);
> ```

BTW This seems dumb to require all calling code to do.  Add a TODO?

---

*[drake/automotive/car_sim_lcm.cc, line 244 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXM1MYp8s1z5UBUOzb:-KVXM1MYp8s1z5UBUOzc:b-i9egz0) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L244)):*
> ```cpp
>       controller->GetMutableSubsystemContext(controller_context,
>                                              rigid_body_plant);
>   rigid_body_plant->SetZeroConfiguration(plant_context);
> ```

BTW This seems dumb to require all calling code to do.  Add a TODO?

---

*[drake/automotive/car_sim_lcm.cc, line 257 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXG7mPmAA12Wav7hIQ:-KVXG7mQ6VMUjzVvgc9d:b-4qmwfw) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L257)):*
> ```cpp
> int main(int argc, char* argv[]) {
>   gflags::ParseCommandLineFlags(&argc, &argv, true);
>   return drake::automotive::main();
> ```

BTW Consider adding HandleSpdlogGflags from https://github.com/RobotLocomotion/drake/blob/master/drake/common/text_logging_gflags.h so that users can turn on verbose logging if/when desired.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
-@sammy-tri +@ggould-tri  (we switched platform review days)

---

Review status: all files reviewed at latest revision, 19 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-:-KV_RCWT0spn55z_6daS:b-w2rqoc)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 3 of 6 files at r1.
Review status: 4 of 6 files reviewed at latest revision, 20 unresolved discussions.

---

*[drake/automotive/automotive_common.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXD9LzxaTj70lEy57Y:-KVa0wj0DEDBk1AFGXKF:b-dt1wjz) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/automotive_common.cc#L12)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I'm not sure this comment is adding value anymore?
</details>
htmlcsscolor.com calls this color "desert sand" :)

---

*[drake/automotive/car_sim_lcm.cc, line 52 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KV_WjT2eTp4LbKupSTd:-KV_WjT2eTp4LbKupSTe:bmivi7x) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L52)):*
> ```cpp
>       systems::plants::joints::kQuaternion, nullptr /* weld to frame */,
>       tree.get());
>   AddFlatTerrainToWorld(tree.get());
> ```

FYI, I believe we are *way* past rule-of-three on near copy-pastes of this stanza.  I don't think that needs to be addressed in this PR, but I have added it to my technical debt punchlist and may blow up this code in the future.

---

*[drake/automotive/car_sim_lcm.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXHRVD-Hhwocb7A11k:-KV_XxL4lPkKvvQJraPp:b-n7ml2c) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L124)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Is this truly the best way to read out the positions and velocities from the tree?  Can't we look up a RigidBody and as it its index?  Or at worst, DEMAND that these hard-coded indicies match up with string Body names from the tree?
</details>
I concur strongly.  Depending on the exact indexes that will result from the parse is brittle at best and an abuse of undocumented parser semantics at worst.  I think @sammy-tri 's recent work on the iiwa shows a good compromise in verifying the values before proceeding.

---

*[drake/automotive/car_sim_lcm.cc, line 213 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXLKarxX2m6dLZW7rE:-KVa2sax19unY8ToM7dr:b-6r8v63) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L213)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Does the controller offer named port getters, or at least kConstant for these magic 0's and 1's here?
</details>
BTW -- +1 to this comment!  `get_input_port(1)` is an odd way to spell "feedback input"

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 18 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Review status: 4 of 6 files reviewed at latest revision, 20 unresolved discussions.

---

*[drake/automotive/automotive_common.cc, line 12 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXD9LzxaTj70lEy57Y:-KVa36dc_f3DeoY2bLft:b4855pb) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/automotive_common.cc#L12)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> htmlcsscolor.com calls this color "desert sand" :)
</details>
Thanks! Done.

---

*[drake/automotive/automotive_common.h, line 26 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXCdm8ApxD0qF9-N0H:-KVa1hU9znkBb0cJNVVs:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/automotive_common.h#L26)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I think this entire overview would be clearer if everything after the first paragraph was replaced with "Two opposite corners of the resulting axis-aligned box are (size/2,size/2,0) and (-size/2,-size/,-depth)" or whatever they are.  I found it difficult to try to parse out the long-form explanation.
</details>
Done.

---

*[drake/automotive/automotive_common.h, line 31 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXDMQzULyVl38RSuqO:-KVa1lpclq0UrB1atADk:bl00cea) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/automotive_common.h#L31)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I think having a pile of free functions in the module-global namespace will _eventually_ be difficult to understand / maintain, but for now it's a fine step along the path to a larger codebase.  Once this file has a handful of functions and users, we can work on refactoring it given the evidence of whatever needs it ends up meeting.
</details>
OK SGTM.

---

*[drake/automotive/car_sim_lcm.cc, line 46 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXG_M2nY6rLHJEoQta:-KVa3HpcluSibMBn7vSA:btzf6vo) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L46)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW I'm not entirely sure we should be using this term (MBD) until the referenced code actually does too.  Whatever, I guess it's just a comment.
</details>
Done. I removed the questionable phrase.

---

*[drake/automotive/car_sim_lcm.cc, line 52 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KV_WjT2eTp4LbKupSTd:-KVaFmn6g2hlb9hBT2U5:b-jq9pa0) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L52)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> FYI, I believe we are *way* past rule-of-three on near copy-pastes of this stanza.  I don't think that needs to be addressed in this PR, but I have added it to my technical debt punchlist and may blow up this code in the future.
</details>
Yes, I agree. We should probably lump this with @amcastro-tri's effort updating `RigidBodyTree`.

---

*[drake/automotive/car_sim_lcm.cc, line 60 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXGyxrjR8itMjtdH9K:-KVaFyMvzK7RTo47qkO5:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L60)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Please document the units for each of the three items.
</details>
Done.

---

*[drake/automotive/car_sim_lcm.cc, line 78 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXHBSN4EYmgjAoT-Be:-KVaGWUST_DosvhGbc2L:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L78)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Please document the units for each of the above.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 2 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/automotive/car_sim_lcm.cc, line 2 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVbFCkmgs-k7IXLfG1i:-KVbFCkmgs-k7IXLfG1j:b-n37bih) ([raw file](https://github.com/robotlocomotion/drake/blob/c1d39c90f6e31f870457d57f4b1ccecaa62bf857/drake/automotive/car_sim_lcm.cc#L2)):*
> ```cpp
> #include <iostream>
> #include <gflags/gflags.h>
> ```

Needs blank line between cstdlib include and library include.

---

*[drake/automotive/car_sim_lcm.cc, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVbFLZd9__qOPxraNqh:-KVbFLZd9__qOPxraNqi:b-vjqyoj) ([raw file](https://github.com/robotlocomotion/drake/blob/c1d39c90f6e31f870457d57f4b1ccecaa62bf857/drake/automotive/car_sim_lcm.cc#L44)):*
> ```cpp
> 
> // Verifies that the order of rigid body names and actuator names within the
> // provided three are as expected.
> ```

typo s/three/tree/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->

liangfok commented:
Partial checkpoint of address of reviewer comments.

---

Review status: 5 of 6 files reviewed at latest revision, 16 unresolved discussions.

---

*[drake/automotive/car_sim_lcm.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXHRVD-Hhwocb7A11k:-KVb6wtkHrn8g1utwkJH:bmmvfbw) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L124)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> I concur strongly.  Depending on the exact indexes that will result from the parse is brittle at best and an abuse of undocumented parser semantics at worst.  I think @sammy-tri 's recent work on the iiwa shows a good compromise in verifying the values before proceeding.
</details>
I added a method called `VerifyCarSimLcmTree()` that verifies the hard-coded indices are correct.

I also added a `TODO` comment saying `feedback_selector_matrix` should be automatically derived based on `RigidBodyTree`.

---

*[drake/automotive/car_sim_lcm.cc, line 149 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXI8emcGanrds9_tGm:-KVboCX-_9Ba7ku74SHV:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L149)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> It seems ill-advised to start executing our Systems before the DiagramBuilder::Build call happens, at least?
</details>
Done.

---

*[drake/automotive/car_sim_lcm.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXIrC8eQEPpCJivW11:-KVbojoNfe8ASDUHIwNA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L154)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The term "throttle velocity" is a bit ambiguous?  Velocity with respect to what position?  Pedal displacement?  Wheel rotation?  Longitudinal vehicle position?
</details>
Done.

---

*[drake/automotive/car_sim_lcm.cc, line 156 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXJCsU8bIqDTTcS-49:-KVboodBEYvGCb1qErCE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L156)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Remove the word `static`.
> https://google.github.io/styleguide/cppguide.html#Static_and_Global_Variables
</details>
Done.

---

*[drake/automotive/car_sim_lcm.cc, line 179 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXKdCtzQa90svxjku9:-KVb7_GLlO3cTprrHOkZ:b-r94eni) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L179)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Why 20?  What are the units?
</details>
This was a known defect carried over from the System 1.0 version of this demo.

https://github.com/RobotLocomotion/drake/issues/3258#issuecomment-256789159

I went ahead and applied the fix in this PR.

---

*[drake/automotive/car_sim_lcm.cc, line 2 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVbFCkmgs-k7IXLfG1i:-KVbnWpGEOmDchys7A7T:bei5rud) ([raw file](https://github.com/robotlocomotion/drake/blob/c1d39c90f6e31f870457d57f4b1ccecaa62bf857/drake/automotive/car_sim_lcm.cc#L2)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> Needs blank line between cstdlib include and library include.
</details>
Done. I removed `#include <iostream>`.

---

*[drake/automotive/car_sim_lcm.cc, line 44 at r3](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVbFLZd9__qOPxraNqh:-KVbnbomNIa4Pa5x3CkU:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/c1d39c90f6e31f870457d57f4b1ccecaa62bf857/drake/automotive/car_sim_lcm.cc#L44)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> typo s/three/tree/
</details>
Done. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-:-KVdtrdwK4LPXsLGph6y:b-mbpfzt)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[drake/automotive/car_sim_lcm.cc, line 179 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXKdCtzQa90svxjku9:-KVdwtg-pd4MHrFvYzqo:b-tuura0) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L179)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> This was a known defect carried over from the System 1.0 version of this demo.
> 
> https://github.com/RobotLocomotion/drake/issues/3258#issuecomment-256789159
> 
> I went ahead and applied the fix in this PR.
</details>
So much better!  Thanks.

---

*[drake/automotive/car_sim_lcm.cc, line 259 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVdx-9LulUAfug2oCDx:-KVdx-9LulUAfug2oCDy:b-1ytl3s) ([raw file](https://github.com/robotlocomotion/drake/blob/e281d55be96d4c380a4022fa64bf0051a4570c45/drake/automotive/car_sim_lcm.cc#L259)):*
> ```cpp
>   // through using a gain of 1. The user's throttle and brake commands need to
>   // be multiplied by a gain of 1 / kWheelRadius and -1. / kWheelRadius to get
>   // the reference rotational velocities for the left and right wheels,
> ```

BTW Consider adding a TODO to account for different left / right target points based on the vehicle's desired curvature?  (I think we're not doing that yet?)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 3 of 6 files at r1, 1 of 2 files at r3, 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

*[drake/automotive/car_sim_lcm.cc, line 124 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXHRVD-Hhwocb7A11k:-KVeOKAmTJ5nSRP-sDmB:bx4pyl0) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L124)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> I added a method called `VerifyCarSimLcmTree()` that verifies the hard-coded indices are correct.
> 
> I also added a `TODO` comment saying `feedback_selector_matrix` should be automatically derived based on `RigidBodyTree`.
</details>
I believe that this verification is adequate until the future rbt rewrite that might get us a good by-name accessor pattern.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I addressed all reviewer comments. PTAL.

---

Review status: 1 of 7 files reviewed at latest revision, 9 unresolved discussions.

---

*[a discussion](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXF6_UHTv70kxHkh_e:-KVeMDqv6nXdpoDqPhhQ:bns6lt3) (no related file):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> (N.B. New discussion thread, don't post PR workflow discussions here.)
> 
> It seems to me like README.md is out of date for this revised program?  I think its fine to _remove_ the incorrect instructions rather than fix them, but I don't think we should post instructions that don't work.  In particular, I think the list of SDFs on the command-line is no longer supported?
</details>
You're right. I updated the README.

---

*[drake/automotive/car_sim_lcm.cc, line 200 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXKjP23XCNNmBJ6mQb:-KVeDEE1H7bF6Dlq5IQh:b4k6ets) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L200)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> To be more obviously failsafe, assert that the matrix size declared by `mimo_gain` was exactly populated by the above left-shift operator.  (Or does Eigen do that automatically?)
</details>
Done. I believe this is guaranteed by Eigen but I added the check anyway.

https://eigen.tuxfamily.org/dox/group__TutorialAdvancedInitialization.html

---

*[drake/automotive/car_sim_lcm.cc, line 202 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXJU7rIDqiK86Ct3vs:-KVeEAme3k6nEEIXX8zI:b-bnun1f) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L202)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW This is an acceptable approach.  Another, possibly-simpler approach is writing a small custom direct-feedthrough System that takes in `Value<lcmt_driving_command_t>` on its single abstract-valued port and during its `EvalOutput` produces whatever controller `BasicVector` is desired, using C++ assignments and elementary math.  The Gain matrix abstraction isn't fundamentally necessary.  This could turn the giant comment matrix above into named getters and setters, with immediately-relevant units and scale comments.
</details>
Good point. I added a `TODO` comment above the declaration of this `MatrixGain` system.

---

*[drake/automotive/car_sim_lcm.cc, line 213 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXLKarxX2m6dLZW7rE:-KVeEOaqoVQ-xhWKasUf:bglafef) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L213)):*
<details><summary><i>Previously, ggould-tri wrote…</i></summary>
> BTW -- +1 to this comment!  `get_input_port(1)` is an odd way to spell "feedback input"
</details>
Unfortunately no. I harped about these "magic numbers" when System 2.0 was first being used but got push back saying these are not "magic numbers".

See for example:
https://reviewable.io/reviews/robotlocomotion/drake/3281#-KSJyzEnf8EZSXYnbYQx

---

*[drake/automotive/car_sim_lcm.cc, line 235 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXLyck7T7lWEpHyp0I:-KVeFp15XMl2ubsU7uHT:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L235)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW This seems dumb to require all calling code to do.  Add a TODO?
</details>
Done.

---

*[drake/automotive/car_sim_lcm.cc, line 244 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXM1MYp8s1z5UBUOzb:-KVeG3NwMY2u-RapCrwE:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L244)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW This seems dumb to require all calling code to do.  Add a TODO?
</details>
Done.

---

*[drake/automotive/car_sim_lcm.cc, line 257 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXG7mPmAA12Wav7hIQ:-KVeGcmYQcqy_1NtGeiF:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L257)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider adding HandleSpdlogGflags from https://github.com/RobotLocomotion/drake/blob/master/drake/common/text_logging_gflags.h so that users can turn on verbose logging if/when desired.
</details>
Done.

---

*[drake/automotive/car_sim_lcm.cc, line 259 at r4](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVdx-9LulUAfug2oCDx:-KVeCERLM_o-EQ-9AqFw:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e281d55be96d4c380a4022fa64bf0051a4570c45/drake/automotive/car_sim_lcm.cc#L259)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW Consider adding a TODO to account for different left / right target points based on the vehicle's desired curvature?  (I think we're not doing that yet?)
</details>
Done.

---

*[drake/automotive/car_simulation.cc, line 154 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXE5E0b0_qlAlp0Lvn:-KVeGyjJ0Ven1BoW-f_B:bj0mf57) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_simulation.cc#L154)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> It looks like this method duplicates the same method in automotive_common?
</details>
Oops. Done. (And found a namespace bug in automotive_common.cc!)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-:-KVeSWY-U0cNtV6TzGbS:b-1edt96)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-:-KVeTMBtY28KviEF4Qz4:bnfp4nl)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 6 of 6 files at r5.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[drake/automotive/car_sim_lcm.cc, line 213 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXLKarxX2m6dLZW7rE:-KVeZd2zjQcqovHTtT8b:b-ohx9pd) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L213)):*
<details><summary><i>Previously, liangfok (Chien-Liang Fok) wrote…</i></summary>
> Unfortunately no. I harped about these "magic numbers" when System 2.0 was first being used but got push back saying these are not "magic numbers".
> 
> See for example:
> https://reviewable.io/reviews/robotlocomotion/drake/3281#-KSJyzEnf8EZSXYnbYQx
</details>
The linked example of "Port numbers for a mux/demux system that map numerically to the vector indices" is a very different argument than "semantically distinct PID controller ports".

---

*[drake/automotive/car_sim_lcm.cc, line 206 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVe_DQVYcOkNJo9kASY:-KVe_DQVYcOkNJo9kASZ:b94b90y) ([raw file](https://github.com/robotlocomotion/drake/blob/c35e036f14d02ce514dcff34ac15c8626b0ba93c/drake/automotive/car_sim_lcm.cc#L206)):*
> ```cpp
>   // [steering angle position, throttle speed, brake speed]
>   //
>   // The throttle and brake speeds are with respect to the vehicl's
> ```

BTW nit Typo `vehicl`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-:-KVe_Mm3z0dkEuMtjsar:bnfp4nl)*
<!-- Sent from Reviewable.io -->

liangfok commented:
I address all reviewer comments. Now waiting on CI.

---

Review status: 6 of 7 files reviewed at latest revision, 3 unresolved discussions.

---

*[drake/automotive/car_sim_lcm.cc, line 213 at r1](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVXLKarxX2m6dLZW7rE:-KVeoeHWBJWNH05Utfxn:bxypxw4) ([raw file](https://github.com/robotlocomotion/drake/blob/0be3ab140e33414dbddec164eb1ccdedebb834db/drake/automotive/car_sim_lcm.cc#L213)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> The linked example of "Port numbers for a mux/demux system that map numerically to the vector indices" is a very different argument than "semantically distinct PID controller ports".
</details>
Done. I added kConstants and a TODO.

---

*[drake/automotive/car_sim_lcm.cc, line 206 at r5](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-KVe_DQVYcOkNJo9kASY:-KVeoxm_-OeCQaX4O--W:boyexbg) ([raw file](https://github.com/robotlocomotion/drake/blob/c35e036f14d02ce514dcff34ac15c8626b0ba93c/drake/automotive/car_sim_lcm.cc#L206)):*
<details><summary><i>Previously, jwnimmer-tri (Jeremy Nimmer) wrote…</i></summary>
> BTW nit Typo `vehicl`.
</details>
Done. Thanks.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886#-:-KVeoz2NE26L-lDmxAGc:b-ufn63m)*
<!-- Sent from Reviewable.io -->

ggould-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->

jwnimmer-tri commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/robotlocomotion/drake/3886)*
<!-- Sent from Reviewable.io -->
